### PR TITLE
Audit for `write_bytes`

### PR DESCRIPTION
**Pull request by adpaco-aws**
### Description of changes: 

Moves and updates the `write_bytes` intrinsic implementation with an alignment check that it was missing. Adds 3 tests cases (2 of them in `expected`).

### Resolved issues:

Part of #727 

### Testing:

* How is this change tested? Adds 3 tests.

* Is this a refactor change? No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 1]
```diff
@@ -1156,4 +1149,34 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        let dst = fargs.remove(0).cast_to(Type::void_pointer());
+        let val = fargs.remove(0).cast_to(Type::c_int());
+        let count = fargs.remove(0);
+        let ty = self.monomorphize(instance.substs.type_at(0));
+        let align = self.is_aligned(ty, dst.clone());
+        let align_check = self.codegen_assert(
+            align,
+            PropertyClass::DefaultAssertion,
+            "`dst` is properly aligned",
+            loc.clone(),
```
Comment (danielsn) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1175]: No need to clone locations

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 2]
```diff
@@ -1156,4 +1149,34 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        );
+        let layout = self.layout_of(ty);
+        let sz = Expr::int_constant(layout.size.bytes(), Type::size_t());
+        let memset_call = BuiltinFn::Memset.call(vec![dst, val, count.mul(sz)], loc);
```
Comment (danielsn) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1179]: We have overflow checking on, right? So we'll see if this overflows?

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 3]
```diff
@@ -1156,4 +1149,34 @@ impl<'tcx> GotocCtx<'tcx> {
         let expr = dst.dereference().assign(src, loc.clone());
         Stmt::block(vec![align_check, expr], loc)
     }
+
+    /// Sets `count * size_of::<T>()` bytes of memory starting at `dst` to `val`
```
Comment (danielsn) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1153]: Like the callout to the documentation

File: tests/expected/intrinsics/write_bytes/out-of-bounds/expected [hunk 4]
```diff
@@ -0,0 +1,2 @@
+FAILURE\
+memset destination region writeable
```
Comment (danielsn) [inline tests/expected/intrinsics/write_bytes/out-of-bounds/expected:2]: newline?

**Review by danielsn — APPROVED** [review]
Looks good, a couple comments

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs (hunk 1 above)
Comment (adpaco-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs]: Thanks, fixed!

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs (hunk 2 above)
Comment (adpaco-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs]: I checked it but we were not detecting overflows here. Updated to include an explicit overflow check for this computation.

File: tests/expected/intrinsics/write_bytes/out-of-bounds/expected (hunk 4 above)
Comment (adpaco-aws) [reply tests/expected/intrinsics/write_bytes/out-of-bounds/expected:2]: I don't include newlines in these because the `expected` testing suite is space-sensitive: `compiletest` would actually check if there's an empty line in the output if we add it.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 5]
```diff
@@ -1156,4 +1149,48 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        // Check that computing `bytes` would not overflow
+        let layout = self.layout_of(ty);
+        let size = Expr::int_constant(layout.size.bytes(), Type::size_t());
+        let bytes = count.mul_overflow(size);
+        let overflow_check = self.codegen_assert(
+            bytes.overflowed.not(),
+            PropertyClass::ArithmeticOverflow,
+            format!("{}: attempt to compute `bytes` which would overflow", intrinsic).as_str(),
+            loc,
+        );
```
Comment (danielsn) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1191]: Not for this PR, but we've done this check > 1 times, which means a helper fn might be good

**Review by danielsn — APPROVED** [review]
